### ISSUE TITLE
api test suite

### ISSUE THREAD

**Issue by vshagur** [openbrewerydb/openbrewerydb-rails-api#40]
Hello. 
I train the skill of writing tests for api. Api.openbrewerydb.org was chosen as an api for tests. You may be interested to know the test results. Or you can apply these tests to improve api. Link to the project with tests [https://github.com/vshagur/openbrewerydb-api-tests]( https://github.com/vshagur/openbrewerydb-api-tests). Work on the tests is not finished yet. If you want to add checks to the test suite, please write which ones. I will add.
Sorry for my English.

**Comment by chrisjm** [comment]
Thank you for contributing, @vshagur ! 

I'll take a deeper look into and run the tests soon. Looks like you have decent coverage, but if you need more checks, you can review the included API unit tests here: 

https://github.com/chrisjm/openbrewerydb-rails-api/tree/master/spec

Great work and keep me updated on when tests are finished! 🍻

**Comment by vshagur** [comment]
Hello. I planned to make the separation of tests into levels according to the number of test data:
1. set of prepared test data for smoke testing
2. a set of generated random data prepared on the basis of a database dump for more complete testing
I needed to check the validity of the data in the database dump (breweries.csv) to be sure of the test results. For this, I wrote a test suite. According to the test results, questions arose. First about the test results.
1. there is no record in the database with id 1863
2. the tag_list field does not exist in the database
3. for records from id 1766 to 1782 in the "state" field, the value is indicated - "District of Columbia" instead of "Columbia"
4. record with id 4506 - website address - 'Http://www.ColfaxAleCellar.com' (the test fell because of the name of the protocol “Http” instead of “http”), also in the database there are names of sites not in lower case
5. records with id 5343 and 3218 - the value "postal_code" contains 4 digits instead of 5. Records with id 4631 and 3635 contain the values ​​- 'TBD' and '48895-97ND'. In addition, in 86 entries the code format is “xxxxx-”, I don’t know how much the minus sign is allowed at the end of the number, in any case, the codes could be brought to a single form.
6. 26 tests fell on checking the “phone” field. The number of characters in the room differed from the standard (9-10).
Now questions.
1. Which fields are required? Some records have fields with empty data. This is postal_code, phone, website_url, city, latitude, longitude, state, street. If the city can be omitted, then the state must be indicated. Valid options are state or city. Similarly for contacts - phone or address. Clarity in this matter will allow us to write more accurate tests.
2. What to do with non-valid data? There are two ways. Correct data or not include it in test cases.
3. There is a problem with cities that may have several valid names. This can complicate the database search.
I understand that the ideal is not attainable. Mistakes have always been, are and will be. Bug fixing takes time. This is your project, so you decide. Please tell me which way you choose. I will add tests based on your answer. The test report can be downloaded [here.](https://drive.google.com/file/d/18s0rk_-aCuyqFPCJhkTukza0HNYoXLC0/view?usp=sharing)

**Comment by vshagur** [comment]
Thank you for the answer. 
Now there is an understanding of which fields are considered mandatory.

Why am I testing api data? I will explain. I have two tasks: 
1. To train the skill of writing tests for api. Therefore, I am experimenting with all possible verification options. I try to cover as many cases as possible and increase coverage. 
2. Get answers to questions - how difficult is it to implement a multi-level test suite? Is there any real benefit from such an implementation? Under the levels you need to understand the different sets of test data that are used in the tests. The choice of level should be determined by the parameters of the cli. Test data is generated depending on the level based on the database dump (for example: level 1 - the minimum set by default, level 2 - the extended set, generation of random data, their number is much larger than in level 1, level 3 - almost full coverage, heavy, long tests).

Generating data for tests (filtering, searching) requires a data source (database dump). It also depends on the quality of the data source. For this reason, I wrote tests for data in a database dump. They need to be considered as an additional tool, and no more. Perhaps you were misled by the poor structure of the project, when everything is piled up (tests, data, additional tools).
You may find these tests useful for solving the "Github Action> Validate Data" task [here](https://github.com/openbrewerydb/openbrewerydb). Problems with maintaining tests should not arise if the data source for generating test data (database dumps) is kept up to date.

**Comment by vshagur** [comment]
Нi, @chrisjm! I finished writing tests for api. You can see the new test results [here](https://github.com/vshagur/openbrewerydb-api-tests/tree/master/reports).

**Comment by chrisjm** [comment]
Hi, @vshagur ! Cool! 

Do you think you could host those HTML files and assets somewhere (Github.io, Netlify, etc) so that we can refer to those test results as we create issues to fix things? Thank you for all of your work!

**Comment by vshagur** [comment]
Done! https://vshagur.github.io/openbrewerydb-api-tests.github.io/index.html
So far, only links, without descriptions and other beauty.

**Comment by vshagur** [comment]
Hello. I collected all the questions that appeared while working on tests in one place. I would be glad if you comment. 
https://github.com/vshagur/openbrewerydb-api-tests.github.io/wiki/Questions.

**Comment by chrisjm** [comment]
@vshagur Thank you for compiling those questions. I don't have time to answer all of the questions right now, but I quickly perused them and most of the questions seem to fall into one of two buckets:

1. **The dataset isn't clean.** Your work will help to correct that. It's not cleaned because I scraped it from a website (https://www.brewersassociation.org/) and immediately imported it into a database. However, from now on I want to only make updates via the official dataset.

2. **Bugs and limited resources.** I started Open Brewery DB as a side-project and have been mostly working on it myself for the past few years. Thanks to you and the rest of the OBDB community, that looks to be changing (yay! thank you!). 

I hope that helps and thanks again for contributing to Open Brewery DB!
